relay-general/src/pii/attachments.rs

```diff
@@ -269,75 +267,58 @@ impl StringMods for [u8] {
-        // this so we need to transmute the lifetimes of the slices we return with
-        // std::slice::from_raw_parts_mut().
-        let mut decoded = String::with_capacity(self.data.len() - self.offset);
-
         loop {
             if self.offset >= self.data.len() {
                 return None;
             }
 
-            decoded.clear();
-            let start = self.offset;
-            let (unprocessed_offset, err) =
-                self.decoder.raw_feed(&self.data[start..], &mut decoded);
-            let end = start + unprocessed_offset;
-
-            if let Some(err) = err {
-                if err.upto > 0 {
-                    self.offset += err.upto as usize;
-                } else {
-                    // This should never happen, but if it does, re-set the decoder and skip
-                    // forward to the next 2 bytes.
-                    self.offset += std::mem::size_of::<u16>(); // TODO: encoding-neutral?!?
-                    self.decoder = self.decoder.from_self();
-                }
-                if decoded.len() > 2 {
-                    let slice = &mut self.data[start..end];
-                    let len = slice.len();
-                    let ptr = slice.as_mut_ptr();
-                    let encoded = unsafe {
-                        WStr::from_utf16le_unchecked_mut(std::slice::from_raw_parts_mut(ptr, len))
-                    };
-                    return Some(WstrSegment { encoded, decoded });
-                } else {
-                    continue;
+            let slice: &mut [u8] = match WStr::from_utf16le_mut(&mut self.data[self.offset..]) {
+                Ok(wstr) => {
+                    self.offset += wstr.len();
+                    unsafe { wstr.as_bytes_mut() }
                 }
-            } else {
-                self.offset += unprocessed_offset;
-                if decoded.len() >= MIN_STRING_LEN {
-                    let slice = &mut self.data[start..end];
-                    let len = slice.len();
-                    let ptr = slice.as_mut_ptr();
-                    let encoded = unsafe {
-                        WStr::from_utf16le_unchecked_mut(std::slice::from_raw_parts_mut(ptr, len))
-                    };
-                    return Some(WstrSegment { encoded, decoded });
-                } else {
-                    return None;
+                Err(err) => {
+                    let start = self.offset;
+                    match err.error_len() {
+                        Some(len) => self.offset += err.valid_up_to() + len,
+                        None => self.offset = self.data.len(),
+                    }
+                    &mut self.data[start..start + err.valid_up_to()]
                 }
+            };
+            if slice.len() < MIN_STRING_LEN * std::mem::size_of::<u16>() {
+                // Quick shortcut, actual char count check done below
+                continue;
+            }
```

Comment:
Is this check really worth it? Lines 311-315 are so cheap that I wouldn't even bother performing this check.

Reply:
yeah, i wondered this too.  more importantly the `.count()` below is too expensive, fixed that as well.